clash-vexriscv/src/VexRiscv/Reset.hs

```diff
@@ -0,0 +1,121 @@
+import qualified Clash.Explicit.Prelude as CE
+
+-- | Data type that contains a reset that is asserted for at least @n@ cycles.
+data MinCyclesReset (dom :: Domain) (n :: Nat) = MinCyclesReset (Reset dom)
+
+type role MinCyclesReset nominal nominal
+
+-- | Extract the underlying reset from a 'MinCyclesReset', this discards the minimum cycles information.
+fromMinCyclesReset :: MinCyclesReset dom n -> Reset dom
+fromMinCyclesReset (MinCyclesReset rst) = rst
+
+-- | Convert a regular reset into a 'MinCyclesReset' that is asserted for at least @1@ cycle.
+toMinCyclesReset ::
+  forall dom.
+  (KnownDomain dom) =>
+  Reset dom ->
+  MinCyclesReset dom 1
+toMinCyclesReset = MinCyclesReset
+
+{- | Convert a regular reset into a 'MinCyclesReset' that is asserted for at least @n@ cycles.
+-- This function is unsafe because it does not insert any logic to ensure the reset is actually
+-- asserted for the required number of cycles. The caller must ensure that the provided reset meets
+-- this requirement.
+-}
+unsafeToMinCyclesReset ::
+  forall dom n.
+  (KnownDomain dom) =>
+  SNat n ->
+  Reset dom ->
+  MinCyclesReset dom n
+unsafeToMinCyclesReset SNat = MinCyclesReset
+
+{- | Coerces a 'MinCyclesReset' into another 'MinCyclesReset' with a different minimum cycle count.
+If the resulting minimum cycle count is higher than the original, `holdReset` is used to extend the reset duration.
+to match the new minimum cycle count.
+
+If the resulting minimum cycle count is lower than the original, the reset is returned as-is.
+This is safe because the original reset already satisfies the higher minimum cycle requirement.
+-}
+coerce ::
+  forall dom n m.
+  (KnownDomain dom, KnownNat n, KnownNat m) =>
+  Clock dom ->
+  MinCyclesReset dom n ->
+  MinCyclesReset dom m
+coerce clk (MinCyclesReset rst) = MinCyclesReset $ case compareSNat (SNat @(n + 1)) (SNat @m) of
+  SNatLE -> holdReset clk enableGen (SNat @(m - n + 1)) rst
+  SNatGT -> rst
+
+{- | Extends a 'MinCyclesReset' to a new 'MinCyclesReset' with a higher minimum cycle count.
+This is done by using `holdReset` to ensure the reset is asserted for the required number of cycles.
+-}
+extendReset ::
+  forall dom n m.
+  (KnownDomain dom, KnownNat n, KnownNat m, n <= m) =>
+  Clock dom ->
+  MinCyclesReset dom n ->
+  MinCyclesReset dom m
+extendReset = coerce
+
+{- | Converts a 'MinCyclesReset' with a higher minimum cycle count to one with a lower minimum cycle count.
+This does not insert any logic and the resulting reset will still be asserted for the higher minimum cycle count.
+This is safe because the original reset already satisfies the higher minimum cycle requirement.
+-}
+fromExtendedReset ::
+  forall dom n m.
+  (KnownNat n, KnownNat m, n <= m) =>
+  MinCyclesReset dom m ->
+  MinCyclesReset dom n
+fromExtendedReset (MinCyclesReset rst) = MinCyclesReset rst
+
+{- | Combine two 'MinCyclesReset's into one that is asserted for at least
+the maximum of the two input minimum cycles.
+-}
+orReset ::
```

Comment:
If you want: this module already scopes it on resets. So IMO this:

```
MinReset.or
```

is better than

```
MinReset.orReset
```

If you do decide to go with the latter, it should be consistent, so `coerce` should be `coerceReset`.